### ISSUE TITLE
In some cases, a useless temporary variable needs to be used with Annotated

### ISSUE THREAD

**Issue by guillaume-alliander** [microsoft/pyright#7049]
**Describe the bug**
I am not too sure what is actually going on. I think that the code is the best way to describe the issue. The workaround works, so it is fine for now, but I do think there's a bug. 

**Code**

```python
from typing import Annotated, Type

def weird[T](p: tuple[Type[T], T]) -> Type[T]:
    return Annotated[p[0], p[1]] # error:  Expected type expression but received "tuple[type[T@weird], T@weird]"
    
    # Alternatives:
    t = p[0]
    return Annotated[t, p[1]] # workaround, no error
    
    return p[0] # not what I want, but no error
```

**VS Code extension or command-line**
On the command line, but pylance has the same issue. 
> pyright --version                                                                                                                            
> pyright 1.1.347

**Comment by erictraut** [comment]
`typing.Annotated` is a special form intended to be used for type annotations. You're using it here as a regular value expression outside of a type annotation. Can you explain what you're trying to do? Are you trying to use it for runtime type checking purposes?

**Comment by erictraut** [comment]
I've improved the error handling for special forms when they're used as value expressions. I don't think this is going to help your use case though. You're doing something that is not allowed in the type system.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.348, which was just published. It will also be included in a future release of pylance.

**Comment by guillaume-alliander** [comment]
> `typing.Annotated` is a special form intended to be used for type annotations. You're using it here as a regular value expression outside of a type annotation. Can you explain what you're trying to do? Are you trying to use it for runtime type checking purposes?

The code example here was the smallest I could make to reproduce the error, but the second parameter of Annotated in my real code is a Pydantic validator. It might indeed be seen as runtime time checking abuse. 

I am not that surprised that what I am doing is dirty (but it works :) ) what surprised me is that using a temporary variable does change the error message. 

The new message in 1.1.348 is clearer indeed (thanks!) but using a temporary variable still has a different behaviour, i.e. another message appears if the temporary variable is not used, which I think would not have been relevant for pyright.

----

For the record, this is the actual thing I am trying to do:

```python
from typing import Annotated, Type

from pydantic import BaseModel, BeforeValidator

class Default:
    def __class_getitem__[T](cls, params: tuple[Type[T], T]) -> Type[T]:
        return Annotated[params[0], BeforeValidator(lambda f: params[1] if f is None else f)]

class Data(BaseModel):
    a: Default[int, 42]

print(Data(a=None)) # type: ignore
```

I want, in a pydantic model, to enforce a default value if `None` is given (`Field(default=42)` does not work because it applies only if no value is given. In my case, `None` _will be_ given, and I want to change it before the rest of the pydantic validation kicks in to ensure that `a` is always an `int`).

The code is the simplest way I have found. Validators would be the logical way to go, but then a new field needs to be added in  2 places: as attributes (with a `Field()` defining metadata to be read during validation) and in the validator itself. Too much boilerplate for my taste. 

I found this way quite clean, even if I am abusing some typing info.
